### PR TITLE
azure: support standalone (non VMSS) instances

### PR DESCRIPTION
**Pull request by bpineau**
Support instances that are not part of a Virtual Machine Scale Set.

Cilium only assumption with regard to machines running in VMSS was
limited to interfaces discovery. Also listing standard (non-VMSS)
network interfaces from the same Azure ResourceGroup is all we need
for those to work with Cilium.

```release-note
Azure: support non VMSS instances
```

### CODE REVIEW

**Comment by ungureanuvladvictor** [comment]
test-me-please

File: pkg/azure/api/api.go [hunk 1]
```diff
@@ -101,8 +101,52 @@ func deriveStatus(err error) string {
 	return "OK"
 }
 
-// describeNetworkInterfaces lists all Azure Interfaces
+// describeNetworkInterfaces lists all Azure Interfaces in the client's resource group
 func (c *Client) describeNetworkInterfaces(ctx context.Context) ([]network.Interface, error) {
+	networkInterfaces, err := c.describeVMSSNetworkInterfaces(ctx)
```
Comment (ungureanuvladvictor) [inline pkg/azure/api/api.go]: Mind renaming this `vmssNetworkInterfaces` ?

**Review by ungureanuvladvictor — CHANGES_REQUESTED** [review]

**Comment by ungureanuvladvictor** [comment]
I am not sure why:
- https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Kernel/1515/
- https://jenkins.cilium.io/job/Cilium-PR-Runtime-4.9/202/

failed but I do not think it can be related to this PR since we have just unit tests for Azure IPAM at the moment.

**Comment by ungureanuvladvictor** [comment]
@bpineau I think you need to modify https://github.com/cilium/cilium/blob/master/pkg/azure/api/api.go#L310-L360 as well to support VMs and not just VMSS.

**Comment by bpineau** [comment]
ohh thanks @ungureanuvladvictor , that one is new to me. I used to create my instances and vmss with pre-allocated ipConfigurations assuming it wasn't supported by Cilium (reason why this pr passed my test). Super good news to see this! will test and update accordingly.

**Comment by ungureanuvladvictor** [comment]
@bpineau Yep! I had a branch locally long time ago that makes allocation work for normal VMs but I never ended up putting it up.

**Review by tgraf — APPROVED** [review]

**Comment by tgraf** [comment]
The existing code looks good. Will review again once `AssignPrivateIpAddresses()` has been adjusted.

@bpineau If you can refresh this PR in the next 24h, we can still merge it in before the feature freeze of 1.8.

**Comment by bpineau** [comment]
Here it is with updated `AssignPrivateIpAddresses()`. Sorry @tgraf , couldn't make it in time.

File: pkg/azure/api/api.go [hunk 2]
```diff
@@ -358,3 +410,38 @@ func (c *Client) AssignPrivateIpAddresses(ctx context.Context, instanceID, vmssN
...
+
+	ipConfigurations = append(*iface.IPConfigurations, ipConfigurations...)
+	iface.IPConfigurations = &ipConfigurations
+
+	future, err := c.interfaces.CreateOrUpdate(ctx, c.resourceGroup, interfaceName, iface)
+	if err != nil {
+		return fmt.Errorf("unable to update interface %s: %s", interfaceName, err)
+	}
+
+	if err := future.WaitForCompletionRef(context.TODO(), c.interfaces.Client); err != nil {
```
Comment (ungureanuvladvictor) [inline pkg/azure/api/api.go]: Use the provided `ctx` that is passed to `assignPrivateIpAddressesToVM`

File: pkg/azure/api/api.go [hunk 3]
```diff
@@ -358,3 +410,38 @@ func (c *Client) AssignPrivateIpAddresses(ctx context.Context, instanceID, vmssN
 
 	return nil
 }
+
+// assignPrivateIpAddressesToVM assign a private IP to an interface attached to a standalone instance
+func (c *Client) assignPrivateIpAddressesToVM(ctx context.Context, instanceID, vmssName, subnetID, interfaceName string, addresses int) error {
```
Comment (ungureanuvladvictor) [inline pkg/azure/api/api.go]: `vmss` param is unused, please remove it.

Comment (ungureanuvladvictor) [reply pkg/azure/api/api.go]: `instanceID` is unused as well, please remove it.

File: pkg/azure/types/types.go [hunk 4]
```diff
@@ -145,6 +146,13 @@ func (a *AzureInterface) extractIDs() {
 		if len(segs) >= 11 {
 			a.vmID = segs[10]
 		}
+	// Interface from a standalone instance:
+	// //subscriptions/xxx/resourceGroups/yyy/providers/Microsoft.Network/networkInterfaces/iii
+	case strings.Contains(a.ID, "/Microsoft.Network/"):
```
Comment (ungureanuvladvictor) [inline pkg/azure/types/types.go:151]: Mind adding a couple of test cases for this via https://github.com/cilium/cilium/blob/master/pkg/azure/types/types_test.go#L75?

File: pkg/azure/api/api.go [hunk 5]
```diff
@@ -310,6 +354,14 @@ func generateIpConfigName() string {
 // AssignPrivateIpAddresses assigns the IPs to the interface as specified by
 // the interfaceName.
 func (c *Client) AssignPrivateIpAddresses(ctx context.Context, instanceID, vmssName, subnetID, interfaceName string, addresses int) error {
+	if vmssName == "" {
```
Comment (ungureanuvladvictor) [inline pkg/azure/api/api.go]: I'm not sure this is the right place to fork the code and do the `vm` vs. `vmss` logic. `AssignPrivateIpAddresses` gets called via https://github.com/cilium/cilium/blob/master/pkg/azure/ipam/node.go#L141. 

How about we change the internal Azure API shim to be:
- `AssignPrivateIpAddressesVM`
- `AssignPrivateIpAddressesVMSS` -- this one following the implementation on the master branch currently

`AssignPrivateIpAddressesVM` -- to be called in https://github.com/cilium/cilium/blob/master/pkg/azure/ipam/node.go#L135 based on the `vmss` being not `""` in the `iface` object we cast via https://github.com/cilium/cilium/blob/master/pkg/azure/ipam/node.go#L136.

The current interface for the internal Azure shim is at https://github.com/cilium/cilium/blob/master/pkg/azure/ipam/instances.go#L29-L34.

@tgraf wdyt about this proposal?

I think another overhaul we could do is to piggy back just on the `vmID` field from https://github.com/cilium/cilium/blob/master/pkg/azure/types/types.go#L122-L123 if we set this correctly as Azure resource names. I believe that one to have the `vmss` name in it and we just have to write some extraction functions. But this is another story which we could tackle later if the code gets too confusing.

**Review by ungureanuvladvictor — CHANGES_REQUESTED** [review]
@bpineau did you run this in a real Azure environment? Currently we do not have e2e tests for the Azure IPAM.

**Comment by ungureanuvladvictor** [comment]
test-me-please

File: pkg/azure/types/types.go (hunk 4 above)
Comment (bpineau) [reply pkg/azure/types/types.go:151]: Done

File: pkg/azure/api/api.go (hunk 5 above)
Comment (bpineau) [reply pkg/azure/api/api.go]: Yes I also considered that option, and don't have much preference either way.

Chose to keep `AssignPrivateIpAddresses` to remain consistent with / comparable to [`EC2API` interface](https://github.com/cilium/cilium/blob/master/pkg/aws/eni/instances.go#L42) (eg. hints what remains to be implemented for a given cloud provider, and perhaps might be later a base for a common cloud API interface).

Not sure if keeping AWS and Azure APIs similar really matters though?
Let me know what's the preferred way.

**Comment by bpineau** [comment]
@ungureanuvladvictor for the review!

Yes I run it on an Azure cluster, and have temporarily disabled my (terraform based) pre-allocations too, to let Cilium handle them. Just re-tested after review changes.

Speaking of which, if you want to try this: `azure.interface-name` CNI setting is not very adapted to standalone instances and interfaces, since those interface's names are not namespaced by VMSS (or VM) and name must be uniq within the resource group.

Support for `first-interface-index` (like for AWS) would be a nice later addition. I personally intend to continue pre-allocating at (non-VMSS) instance creation due to that, but doing so still requires the first patch (hence the initial PR submission content).

File: pkg/azure/api/api.go (hunk 5 above)
Comment (ungureanuvladvictor) [reply pkg/azure/api/api.go]: I think they are different and we never had the discussion to consolidate into one shim. My preference is to not do that given all the subtle differences between AWS/Azure (and also GCP for the future).

**Review by ungureanuvladvictor — COMMENTED** [review]
`ctx` change + unit tests look good. Let's keep the discussion a bit open on the Azure API change.

@bpineau for `first-interface-index` but how will you know which index the interface is attached at in Azure? I think in Azure the only thing that identifies the interface on the machine is the `name` of the interface.

**Comment by aanm** [comment]
@bpineau are the changes requested by Vlad applied?

File: pkg/azure/api/api.go (hunk 5 above)
Comment (bpineau) [reply pkg/azure/api/api.go]: @ungureanuvladvictor implemented your suggestion

**Comment by bpineau** [comment]
@ungureanuvladvictor updated (and re-tested on a real Azure cluster) to match your suggested layout. 

(unrelated: you're probably right about interfaces order; not sure what a proper way to identify interfaces that can't have uniq names would be).

**Review by ungureanuvladvictor — APPROVED** [review]

**Comment by ungureanuvladvictor** [comment]
test-me-please

**Pull request merged**